controller/twilio.go

```diff
@@ -0,0 +1,186 @@
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"bitbucket.org/ckvist/twilio/twiml"
+	"github.com/gorilla/mux"
+)
+
+// CallbackResponse is the json struct as returned from twilio in the callbacks
+type CallbackResponse struct {
+	UUID    string `json:"uuid"`
+	Message string `json:"message"`
+	Error   string `json:"error"`
+}
+
+// ReceiveSMSReply receives the SMS reply callback from Twilio and deletes the notification if the
+// response text matches the code sent with the original SMS notification
+func (a *Controller) ReceiveSMSReply(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		log.Println("ReceiveSMSReply() r.ParseForm() error:", err)
+		http.Error(w, "", http.StatusInternalServerError)
+		return
+	}
+
+	// We should have a "From" parameter being passed from Twilio
+	recipient := r.FormValue("From")
+	if recipient == "" {
+		log.Println("ReceiveSMSReply() error: 'From' parameter was not provided in response")
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	}
+
+	// Our conversation key is a combination of the recipient's phone number and the 3-digit code
+	// that they sent in reply
+	conversationKey := fmt.Sprint(recipient, "::", r.FormValue("Body"))
+
+	uuid, err := a.n.GetConversation(conversationKey)
+	if err != nil {
+		a.n.SendSMS(recipient, "I'm sorry but I don't recognize that response.   Please acknowledge with the three-digit code from the notfication you received.", "", true)
```

Comment:
This file is another candidate for internationlization.   We'll want to include a way to pass the language code to the Twilio API so that the words are pronounced correctly.